impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java

```diff
@@ -32,13 +41,82 @@
+     * Verifies that each broker in the cluster is returning the expected cluster size.
+     *
+     * @param connectionConfig the connection config
+     * @param timeout the timeout
+     * @param timeUnit the time unit
+     * @param expectedBrokerCount the expected broker count
+     */
+    public static void awaitExpectedBrokerCountInClusterViaTopic(Map<String, Object> connectionConfig, int timeout, TimeUnit timeUnit, Integer expectedBrokerCount) {
+        try (var admin = Admin.create(connectionConfig)) {
+            log.debug("Creating topic: {} via {}", CONSISTENCY_TEST, connectionConfig.get(BOOTSTRAP_SERVERS_CONFIG));
+            admin.createTopics(Set.of(new NewTopic(CONSISTENCY_TEST, expectedBrokerCount, getReplicationFactor(expectedBrokerCount))))
+                    .all()
+                    .whenComplete((unused, throwable) -> {
+                        log.debug("Create topic future completed.");
+                        if (throwable != null) {
+                            log.warn("Failed to create topic: {} due to {}", CONSISTENCY_TEST, throwable.getMessage(), throwable);
+                            Assert.fail("Failed to create topic: " + CONSISTENCY_TEST + "  due to " + throwable.getMessage());
+                        }
+                    });
+            log.debug("Waiting for {} to be replicated to {} brokers", CONSISTENCY_TEST, expectedBrokerCount);
+            awaitCondition(timeout, timeUnit)
+                    .until(() -> {
+                        log.debug("Calling describe topic");
+                        final var promise = new CompletableFuture<Boolean>();
+                        admin.describeTopics(Set.of(CONSISTENCY_TEST))
+                                .allTopicNames()
+                                .whenComplete((topicDescriptions, throwable) -> {
+                                    if (throwable != null) {
+                                        if (throwable instanceof CompletionException && throwable.getCause() instanceof UnknownTopicOrPartitionException) {
+                                            log.debug("Cluster quorum test topic ({}) doesn't exist yet", CONSISTENCY_TEST);
+                                        }
+                                        else {
+                                            log.debug("Unexpected failure describing topic: {} due to {}", CONSISTENCY_TEST, throwable.getMessage(), throwable);
+                                        }
+                                        promise.complete(false);
+                                    }
+                                    else {
+                                        log.debug("Current topicDescriptions: {}", topicDescriptions);
+                                        var topicDescription = topicDescriptions.get(CONSISTENCY_TEST);
+                                        final long partitionLeaders = topicDescription.partitions().stream().map(TopicPartitionInfo::leader)
```

Comment:
I understand your approach now: checking that the number of leaders == number of brokers.  That seems smart.  Should there be a distinct in the pipeline?

Reply:
TL;DR I'm going to switch to RF=expected broker count while keeping the logic to check the result of describeTopic.

I think a `distinct` is a good addition yes.

The partitionLeader pattern comes from the abnormal workloads testing done in a previous life. In which we saw intermittent failures where the topic was created but we couldn't publish to it. Which is noted in the [AdminClient#createTopics javadoc](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/admin/Admin.html#createTopics(java.util.Collection,org.apache.kafka.clients.admin.CreateTopicsOptions)) 
> It may take several seconds after [CreateTopicsResult](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/admin/CreateTopicsResult.html) returns success for all the brokers to become aware that the topics have been created.

The fix in that case was to ensure each partition had a leader. In that testing scenario we cared about the number of partitions and the distribution of them which might be why we ended up doing more than just asserting the number of replicas. In our current case we would still need to check the result of describeTopic.

Poking at the describe topic results suggests the following would also work:
```java
final long partitionLeaders = topicDescription.partitions()
        .stream()
        .map(TopicPartitionInfo::replicas)
        .flatMap(List::stream)
        .filter(Objects::nonNull)
        .distinct()
        .count();
```

You're making a good point about the topic creation failing, we probably need to retry that too.

Reply:
Sounds good.